File: Backend/src/main/java/backend/osiris/entity/product/Product.java
package backend.osiris.entity.product;

import backend.osiris.entity.BaseEntity;
import backend.osiris.entity.client.PreOrder;
import backend.osiris.entity.client.Wish;
import backend.osiris.entity.general.Image;
import backend.osiris.entity.promotion.Promotion;
import backend.osiris.entity.review.Review;
import backend.osiris.utils.JsonNodeConverter;
import com.fasterxml.jackson.annotation.JsonBackReference;
import com.fasterxml.jackson.annotation.JsonManagedReference;
import com.fasterxml.jackson.databind.JsonNode;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.experimental.Accessors;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

@AllArgsConstructor
@NoArgsConstructor
@Getter
@Setter
@Accessors(chain = true)
@Entity
@Table(name = "product")
public class Product extends BaseEntity {
    @Column(name = "name", nullable = false)
    private String name;

    @Column(name = "code", nullable = false, unique = true)
    private String code;

    @Column(name = "slug", nullable = false, unique = true)
    private String slug;

    @Column(name = "short_description")
    private String shortDescription;

    @Column(name = "description", columnDefinition = "TEXT")
    private String description;

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true)
    @JsonManagedReference
    private List<Image> images = new ArrayList<>();

    @Column(name = "status", nullable = false, columnDefinition = "TINYINT")
    private Integer status;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "category_id")
    @JsonBackReference
    private Category category;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "brand_id")
    @JsonBackReference
    private Brand brand;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "supplier_id")
    @JsonBackReference
    private Supplier supplier;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "unit_id")
    @JsonBackReference
    private Unit unit;

    @ManyToMany(cascade = {CascadeType.PERSIST, CascadeType.MERGE})
    @JoinTable(
            name = "product_tag",
            joinColumns = @JoinColumn(name = "product_id", nullable = false),
            inverseJoinColumns = @JoinColumn(name = "tag_id", nullable = false)
    )
    private Set<Tag> tags = new HashSet<>();

    @Column(name = "specifications", columnDefinition = "JSON")
    @Convert(converter = JsonNodeConverter.class)
    private JsonNode specifications;

    @Column(name = "properties", columnDefinition = "JSON")
    @Convert(converter = JsonNodeConverter.class)
    private JsonNode properties;

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true)
    @JsonManagedReference
    private List<Variant> variants = new ArrayList<>();

    @Column(name = "weight")
    private Double weight;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "guarantee_id")
    @JsonBackReference
    private Guarantee guarantee;

    @OneToMany(mappedBy = "product")
    private List<Wish> wishes = new ArrayList<>();

    @OneToMany(mappedBy = "product")
    private List<PreOrder> preorders = new ArrayList<>();

    @OneToMany(mappedBy = "product")
    private List<Review> reviews = new ArrayList<>();

    @ManyToMany(mappedBy = "products")
    private Set<Promotion> promotions = new HashSet<>();
}

File: Frontend/src/models/Product.ts
import { CollectionWrapper } from "@/datas/CollectionWrapper";
import BaseResponse from "./BaseResponse";
import { BrandResponse } from "./Brand";
import { GuaranteeResponse } from "./Guarantee";
import { ImageResponse, ImageRequest } from "./Image";
import { SupplierResponse } from "./Supplier";
import { TagResponse } from "./Tag";
import { UnitResponse } from "./Unit";
import { VariantPropertyItem, VariantRequest } from "./Variant";

export interface ProductResponse extends BaseResponse {
    name: string;
    code: string;
    slug: string;
    shortDescription: string | null;
    description: string | null;
    images: ImageResponse[];
    status: number;
    category: CategoryResponse | null;
    brand: BrandResponse | null;
    supplier: SupplierResponse | null;
    unit: UnitResponse | null;
    tags: TagResponse[];
    specifications: CollectionWrapper<SpecificationItem> | null;
    properties: CollectionWrapper<ProductPropertyItem> | null;
    variants: ProductResponse_VariantResponse[];
    weight: number | null;
    guarantee: GuaranteeResponse | null;
}

interface CategoryResponse extends BaseResponse {
    name: string;
    slug: string;
    description: string | null;
    thumbnail: string | null;
    status: number;
}

export interface ProductResponse_VariantResponse extends BaseResponse {
    sku: string;
    cost: number;
    price: number;
    properties: CollectionWrapper<VariantPropertyItem> | null;
    status: number;
}

export interface SpecificationItem {
    id: number;
    name: string;
    code: string;
    value: string;
}

export interface ProductPropertyItem {
    id: number;
    name: string;
    code: string;
    value: string[];
}

export interface ProductRequest {
    name: string;
    code: string;
    slug: string;
    shortDescription: string | null;
    description: string | null;
    images: ImageRequest[];
    status: number;
    categoryId: number | null;
    brandId: number | null;
    supplierId: number | null;
    unitId: number | null;
    tags: ProductRequest_TagRequest[];
    specifications: CollectionWrapper<SpecificationItem> | null;
    properties: CollectionWrapper<ProductPropertyItem> | null;
    variants: VariantRequest[];
    weight: number | null;
    guaranteeId: number | null;
}

export interface ProductRequest_TagRequest {
    id?: number;
    name?: string;
    slug?: string;
    status?: number;
}
